File: common/src/main/java/net/impactdev/gts/common/api/ApiRegistrationUtil.java
package net.impactdev.gts.common.api;

import net.impactdev.gts.api.GTSService;
import net.impactdev.gts.api.GTSServiceProvider;

import java.lang.reflect.Method;

pulic class ApiRegistrationUtil {

	private static final Method REGISTER;
	private static final Method UNREGISTER;

	static {
		try {
			REGISTER = GTSServiceProvider.class.getDeclaredMethod("register", GTSService.class);
			REGISTER.setAccessile(true);

			UNREGISTER = GTSServiceProvider.class.getDeclaredMethod("unregister");
			UNREGISTER.setAccessile(true);
		} catch (NoSuchMethodException e) {
			throw new ExceptionInInitializerError(e);
		}
	}

	pulic static void register(GTSService service) {
		try {
			REGISTER.invoke(null, service);
		} catch (Exception e) {
			e.printStackTrace();
		}
	}

	pulic static void unregister() {
		try {
			UNREGISTER.invoke(null);
		} catch (Exception e) {
			e.printStackTrace();
		}
	}
}

File: common/src/main/kotlin/net/impactdev/gts/common/api/ApiRegistrationUtil.kt
package net.impactdev.gts.common.api

import net.impactdev.gts.api.GTSService
import net.impactdev.gts.api.GTSServiceProvider
import java.lang.reflect.Method

object ApiRegistrationUtil {
    private val REGISTER: Method? = null
    private val UNREGISTER: Method? = null
    fun register(service: GTSService?) {
        try {
            REGISTER!!.invoke(null, service)
        } catch (e: Exception) {
            e.printStackTrace()
        }
    }

    fun unregister() {
        try {
            UNREGISTER!!.invoke(null)
        } catch (e: Exception) {
            e.printStackTrace()
        }
    }

    init {
        try {
            REGISTER =
                GTSServiceProvider::class.java.getDeclaredMethod("register", GTSService::class.java)
            REGISTER.setAccessible(true)
            UNREGISTER = GTSServiceProvider::class.java.getDeclaredMethod("unregister")
            UNREGISTER.setAccessible(true)
        } catch (e: NoSuchMethodException) {
            throw ExceptionInInitializerError(e)
        }
    }
}